packages/expo-localization/ios/EXLocalization/EXLocalization.m

```diff
@@ -20,32 +20,43 @@ @implementation EXLocalization
 
 - (NSDictionary *)constantsToExport
 {
-  NSArray<NSString *> *preferredLocales = [NSLocale preferredLanguages];
-  if (![preferredLocales count]) {
-    NSString *currentLocale = [[NSLocale currentLocale] localeIdentifier];
-    if (currentLocale == nil) {
-      currentLocale = @"en_US";
-    }
-    preferredLocales = @[currentLocale];
+  NSLocale *locale = [NSLocale currentLocale];
+  NSString *languageCode = [locale objectForKey:NSLocaleLanguageCode];
+  NSArray<NSString *> *languageIds = [NSLocale preferredLanguages];
+  if (![languageIds count]) {
+    languageIds = @[@"en-US"];
   }
   
-  NSTimeZone *currentTimeZone = [NSTimeZone localTimeZone];
-  NSString *region = [[NSLocale currentLocale] objectForKey:NSLocaleCountryCode];
-  
   return @{
-           @"isRTL": @([self isRTL]),
-           @"locale": [preferredLocales objectAtIndex:0],
-           @"locales": preferredLocales,
-           @"timezone": [currentTimeZone name],
-           @"isoCurrencyCodes": [NSLocale ISOCurrencyCodes],
-           @"region": UMNullIfNil(region)
-           };
+    @"currency": [EXLocalization currencyCodeForLocale:locale] ?: @"USD",
+    @"decimalSeparator": [locale objectForKey:NSLocaleDecimalSeparator] ?: @".",
+    @"groupingSeparator": [locale objectForKey:NSLocaleGroupingSeparator] ?: @",",
+    @"isoCurrencyCodes": [NSLocale ISOCurrencyCodes],
+    @"isMetric": @([[locale objectForKey:NSLocaleUsesMetricSystem] boolValue]),
+    @"isRTL": @((BOOL)([NSLocale characterDirectionForLanguage:languageCode] == NSLocaleLanguageDirectionRightToLeft)),
+    @"locale": [languageIds objectAtIndex:0],
+    @"locales": languageIds,
```

Comment:
On iOS, Locale IDs and Language IDs are different. A Language ID uses hyphens to separate the base language from the dialect's region ("en-US"). A Locale ID uses underscores to separate the language from the region ("en_US").

https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html#//apple_ref/doc/uid/10000171i-CH15

We can make this breaking change if we need to, but we should be really intentional about it and make sure it's what we want to use going forward. (And document it in the changelog.)

Reply:
Yeah I know. This was hugely confusing to me as well. I was initially expecting underscored locales, but got language tags instead. Also some iOS code paths were returning underscored locales and others were returning language tags.

> It looks like W3C and other standards tend to use hyphens. The W3C unicode locale draft says all locale identifies are valid language tags, even. So I think what you have in this PR (namely, using BCP 47) is correct.

Yep that was my conclusion as well. Language-tags seem more consistent and in some cases convey more information (especially on iOS when scripts and variants are involved which are not encoded in the locale-id). The API naming of `locale` and `locales` seem unfortunate here though. It would probably have been better if they were called `language` and `languages`. It would be possible to add the `language` and `languages` props, but we'd have to be very careful and deliberate in changing the behaviour of the `locale` property. Also, getting the locale-id is not possible on web (although you could try to reformat the language-id), and it's not possible to get a list of actual "preferred" locales on your native device.

I personally see two options here:

### Option A

- Add `language` property (same behaviour as current `locale` but just renamed)
- Add `languages` property (same behaviour as current `locales` but just renamed)
- Add `localeId` property (new property that returns the actual underscored locale-id from native)
- Deprecate `locale` property (but still works as before)
- Deprecate `locales` property (but still works as before)
- Do it in a JavaScript compatible manner when possible

### Option B

- Leave it as it is now and document clearly

What do you think?